### PR TITLE
Chess-Results.com Notifier

### PR DESCRIPTION
**Pull request by danylykkk**

### CODE REVIEW

**Comment by danylykkk** [comment]
SMTP Server
image

**Review by polyakovartem — CHANGES_REQUESTED** [review]
Please make clear subscription API for player and tournament

**Comment by danylykkk** [comment]
You mean my own API?)

**Comment by polyakovartem** [comment]
Let me rephrase it. I don’t understand how to use your application to subscribe for tournament participant update nor for tournament update

File: data.dv [hunk 1]
```diff
@@ -0,0 +1 @@
...rntner JUGENDMASTER Gruppe A ", "data": [[null, null], ["32", "Wa\u00dfmann Moritz"]]}, {"link": "tnr500469.aspx?lan=1", "title": "Campeonato Provincial Superior Absoluto de Ajedrez Catamarca 2019 ", "data": [["tnr500469.aspx?lan=1&art=2&rd=1", "1"], ["27", "Vega Ruben"]]}, {"link": "tnr500492.aspx?lan=1", "title": " Torneo Ciudad de Juliaca ", "data": [["tnr500492.aspx?lan=1&art=2&rd=5", "5"], ["12", "Zegarra MAria"]]}, {"link": "tnr500434.aspx?lan=1", "title": "2019-2020 K\u00dc\u00c7\u00dcKLER SAMSUN \u0130L B\u0130R\u0130NC\u0130L\u0130\u011e\u0130 SATRAN\u00c7 TURNUVASI 8 YA\u015e GENEL (2012 DO\u011eUMLU SPORCULAR) ", "data": [[null, null], ["62", "YILMAZ AHMET"]]}, {"link": "tnr467542.aspx?lan=1", "title": "2. Wiener Weihnachtsopen 2019 ", "data": [["tnr467542.aspx?lan=1&art=2&rd=6", "6"], ["81", "Seemann Thomas"]]}, {"link": "tnr500303.aspx?lan=1", "title": "XXXII ABIERTO CIUDAD DE ARMILLA  |  TORNEO DE MENORES listado provisional (Eloint es Elo FADA) ", "data": [[null, null], ["31", "ZAMORA FORASTERO Luis"]]}, {"link": "tnr467545.aspx?lan=1", "title": "2. Wiener Weihnachtsopen 2019 Blitzschach inklusive ASV\u00d6-LM ", "data": [[null, null], ["57", "Wallinger Nikolaus"]]}]
```
Comment (polyakovartem) [inline data.dv:1]: what is this file for?

File: Chess.py [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+        self.appending = True
+        self.data_update = False
+        if os.path.isfile("data.dv"):
+            with open('data.dv') as saved_data:
+                self.data = json.load(saved_data)
+
+        self.saveing_thread = Thread(target = self.saveing)
+        self.saveing_thread.start()
+
+    def saveing (self):
```
Comment (polyakovartem) [inline Chess.py:28]: saveing -> saving

File: Chess.py [hunk 3]
```diff
@@ -0,0 +1,216 @@
+from threading import Thread
+import os
+import smtplib
+from email.message import EmailMessage
+import json
+import requests
+import urllib.request
+from bs4 import BeautifulSoup
+
+class EventNotifier:
```
Comment (polyakovartem) [inline Chess.py:10]: please decuple this class from data processing/parsing, the intention of this class should be event notification no more, no less

File: Chess.py [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+        self.tournaments = {}
+        self.last_updated = ""
+        self.domain = "http://chess-results.com/"
+        self.data = []
+        self.app_run = True
+        self.appending = True
+        self.data_update = False
+        if os.path.isfile("data.dv"):
+            with open('data.dv') as saved_data:
+                self.data = json.load(saved_data)
```
Comment (polyakovartem) [inline Chess.py:23]: this class should not care about saving data, move this logic out

File: Chess.py [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+    def run (self):
+        self.appending = False
+        while self.app_run:
+            #self.get_update()
+            for url in self.participants:
+                self.get_patricipant_update(url)
+            for url in self.tournaments:
+                self.tournament_update(url)
+
+    def append_participant (self, url):
```
Comment (polyakovartem) [inline Chess.py:47]: this function is too big, split it into smaller functions with clean names

File: Chess.py [hunk 6]
```diff
@@ -0,0 +1,216 @@
...
+
+        soup = BeautifulSoup(html_content, "html.parser")
+        data = soup.find_all("table", {"class": "CRs2"})[0]
+        trs = data.find_all("tr")
+        updated = trs[1].find_all("td")[1]
+        if self.last_updated != updated.text:
+            self.last_updated = updated.text
+            self.tournament_update(updated.find_all("a")[0]['href'], self.last_updated)
+
+    def get_patricipant_update (self, url):
```
Comment (polyakovartem) [inline Chess.py:115]: split this function please

File: Chess.py [hunk 7]
```diff
@@ -0,0 +1,216 @@
...
+                    item["data"][1][0] = data[1][0]
+                    item["data"][1][1] = data[1][1]
+                    update = True
+                if update:
+                    self.data_update = True
+                return
+        self.data.append({"link": link, "title": title, "data": data})
+        self.data_update = True
+
+    def send_email (self, message):
```
Comment (polyakovartem) [inline Chess.py:179]: notification object should not be dependent on emails, what if you need to add more notification transports in the future?

File: Chess.py [hunk 8]
```diff
@@ -0,0 +1,216 @@
...
+        msg = EmailMessage()
+        msg.set_content(message)
+        msg['Subject'] = f'Chess updates'
+        msg['From'] = "danylyk@python.com"
+        for email in self.emails:
+            msg['To'] = email
+            s = smtplib.SMTP('localhost', 1025)
```
Comment (polyakovartem) [inline Chess.py:186]: are you sure you need to create a new SMTP connection from every new email?

File: Chess.py [hunk 9]
```diff
@@ -0,0 +1,216 @@
...
+            s.send_message(msg)
+        s.quit()
+    
+    def tournament_update (self, link, title):
```
Comment (polyakovartem) [inline Chess.py:190]: the same comment as for methods from above

File: Client.py [hunk 10]
```diff
@@ -0,0 +1,9 @@
+from Chess import EventNotifier
+
+if __name__ == "__main__":
+    emails = ["webdvitaly@gmail.com", "web.vitaly@rambler.ru"]
+    notifier = EventNotifier(emails)
+    notifier.append_participant("http://chess-results.com/tnr83447.aspx?lan=1&art=9&snr=23")
```
Comment (polyakovartem) [inline Client.py:6]: append_participant -> subscribe_for_participant_updates

File: Client.py [hunk 11]
```diff
@@ -0,0 +1,9 @@
+from Chess import EventNotifier
+
+if __name__ == "__main__":
+    emails = ["webdvitaly@gmail.com", "web.vitaly@rambler.ru"]
+    notifier = EventNotifier(emails)
+    notifier.append_participant("http://chess-results.com/tnr83447.aspx?lan=1&art=9&snr=23")
+    notifier.append_participant("http://chess-results.com/tnr149322.aspx?lan=1&art=9&snr=118")
+    notifier.append_tournament("http://chess-results.com/tnr219320.aspx?lan=1")
```
Comment (polyakovartem) [inline Client.py:8]: append_tournament->subscribe_for_tournament_updates

**Review by polyakovartem — CHANGES_REQUESTED** [review]
in general very good, just few changes needed
